core/src/main/java/com/linecorp/armeria/common/logging/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,351 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import java.beans.BeanInfo;
+import java.beans.IntrospectionException;
+import java.beans.Introspector;
+import java.beans.PropertyDescriptor;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.DecoratingClientFunction;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.RpcResponse;
+import com.linecorp.armeria.common.http.HttpSessionProtocols;
+import com.linecorp.armeria.server.DecoratingServiceFunction;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Gauge;
+import io.prometheus.client.Summary;
+
+/**
+ * Provides request decorator for server and client sides. Decorators will record metrics into
+ * Prometheus {@link CollectorRegistry}.
+ * @param <I> {@link Request} type
+ * @param <O> {@link Response} type
+ */
+public final class PrometheusMetricRequestDecorator<I extends Request, O extends Response>
+        implements DecoratingServiceFunction<I, O>, DecoratingClientFunction<I, O> {
+
+    private static class MetricAdapter {
+        private AtomicReference<MetricAdapter> initialized;
+
+        private Summary timer;
+        private Counter success;
+        private Counter failure;
+        private Gauge activeRequests;
+        private Summary requestBytes;
+        private Summary responseBytes;
+
+        private List<Method> metricLabelValueMethods;
+
+        MetricAdapter() {
+            initialized = new AtomicReference<>();
+        }
+
+        public MetricAdapter initialize(MetricLabel<?> namingFunction,
+                                        PrometheusRegistryWrapper collectorRegistry) {
+            if (initialized.compareAndSet(null, this)) {
+                BeanInfo metricLabelValueBeanInfo;
+
+                try {
+                    metricLabelValueBeanInfo = Introspector.getBeanInfo(namingFunction.getValue().getClass());
+                } catch (IntrospectionException e) {
+                    throw new RuntimeException(e);
+                }
+
+                String[] labels = Arrays.asList(metricLabelValueBeanInfo.getPropertyDescriptors())
+                                        .stream()
+                                        .skip(1)
```

Comment:
Thanks I mean't to fix that. the first property is ``class`` but theres actually no guarantee for ordering. 
Actually it seems to be returning properties alphabetically